# Using Pandas to clean data

## Importing libraries

When initiating a python session or script, import the necessary libraries. Commonly-used libraries have import conventions, such as abbreviating numpy as np. When you call a numpy method in your script, use np.method(). For example, to generate a random number you would use my_number = np.random.randn


In [304]:
from pandas import *
import numpy as np
import matplotlib.pyplot as plt

## Coursera course project

I did this in R for the Coursera class, and am now redoing it in pandas/python to ensure some basic functionality isn't being lost (i.e. that R has critical things we do actually need).  

From the Coursera website: 

>The purpose of this project is to demonstrate your ability to collect, work with, and clean a data set. The goal is to prepare tidy data that can be used for later analysis. You will be graded by your peers on a series of yes/no questions related to the project. You will be required to submit: 1) a tidy data set as described below, 2) a link to a Github repository with your script for performing the analysis, and 3) a code book that describes the variables, the data, and any transformations or work that you performed to clean up the data called CodeBook.md. You should also include a README.md in the repo with your scripts. This repo explains how all of the scripts work and how they are connected. 

>One of the most exciting areas in all of data science right now is wearable computing - see for example this article . Companies like Fitbit, Nike, and Jawbone Up are racing to develop the most advanced algorithms to attract new users. The data linked to from the course website represent data collected from the accelerometers from the Samsung Galaxy S smartphone. A full description is available at the site where the data was obtained:

> http://archive.ics.uci.edu/ml/datasets/Human+Activity+Recognition+Using+Smartphones

>Here are the data for the project:

> https://d396qusza40orc.cloudfront.net/getdata%2Fprojectfiles%2FUCI%20HAR%20Dataset.zip

> You should create one R script called run_analysis.R that does the following. 

>1. Merges the training and the test sets to create one data set.
2. Extracts only the measurements on the mean and standard deviation for each measurement. 
3. Uses descriptive activity names to name the activities in the data set
4. Appropriately labels the data set with descriptive variable names. 
5. From the data set in step 4, creates a second, independent tidy data set with the average of each variable for each activity and each subject.

I have saved the files in the .zip folder to my current directory. We only have to read in a few files: 

First, read in the files that associate each subject (1-30) with an observation (row). The data comes in two parts: training and testing data. After reading the data, we will merge the two sets into one. 

In [305]:
subject_train = read_csv('C:/Users/ejones33/Documents/data/UCI HAR Dataset/train/subject_train.txt',header=None,names=["subject"])
subject_test = read_csv('C:/Users/ejones33/Documents/data/UCI HAR Dataset/test/subject_test.txt',header=None,names=["subject"])

In [306]:
n_train = len(subject_train)
n_test = len(subject_test)
n_train, n_test

(7352, 2947)

We see that there are 7352 training observations and 2947 test observations. Our complete data set will have 10,299 observations. 

In [307]:
subject = subject_train.append(subject_test) 

In [308]:
subject.head()

subject
0        1
1        1
2        1
3        1
4        1

Next, we read in the actual data associated with each observation. There were 561 measurement types recorded, but we only care about the ones with mean or standard deviation in them. There are 66 of these variables. We will select these later.

The data is split into two `.txt` files for both testing and training. The `y_test` and `y_train` files are vectors containing what activity was being performed. The `X_test` and `X_train` files are tables containing each of the 561 measurements for each observation. 

In [309]:
y_train = read_csv("C:/Users/ejones33/Documents/data/UCI HAR Dataset/train/y_train.txt",header=None,names=["activity"])
y_test = read_csv("C:/Users/ejones33/Documents/data/UCI HAR Dataset/test/y_test.txt",header=None,names=["activity"])
y_train.head()

activity
0         5
1         5
2         5
3         5
4         5

In [310]:
features = read_csv("C:/Users/ejones33/Documents/data/UCI HAR Dataset/features.txt",sep=" ",header=None,usecols=[1],names=["features"])
features.head()

features
0  tBodyAcc-mean()-X
1  tBodyAcc-mean()-Y
2  tBodyAcc-mean()-Z
3   tBodyAcc-std()-X
4   tBodyAcc-std()-Y

In [311]:
feature_names = Series(features["features"])
feature_names[:10]

0    tBodyAcc-mean()-X
1    tBodyAcc-mean()-Y
2    tBodyAcc-mean()-Z
3     tBodyAcc-std()-X
4     tBodyAcc-std()-Y
5     tBodyAcc-std()-Z
6     tBodyAcc-mad()-X
7     tBodyAcc-mad()-Y
8     tBodyAcc-mad()-Z
9     tBodyAcc-max()-X
Name: features, dtype: object

In [312]:
substring = '(mean|std)'
substring = 'mean\(\)|std\(\)'
is_mean_std = feature_names.str.contains(substring)
#is_mean_std.index[:10]
is_mean_std[:10]

0     True
1     True
2     True
3     True
4     True
5     True
6    False
7    False
8    False
9    False
Name: features, dtype: bool

In [313]:
my_cols = np.where(is_mean_std)
my_cols = list(my_cols[0])
my_cols[:20]

[0, 1, 2, 3, 4, 5, 40, 41, 42, 43, 44, 45, 80, 81, 82, 83, 84, 85, 120, 121]

In [314]:
feature_names = list(feature_names[is_mean_std])

In [315]:
len(feature_names)

66

In [316]:
X_train = read_fwf("C:/Users/ejones33/Documents/data/UCI HAR Dataset/train/X_train.txt",
                   #sep=" ",
                   skipinitialspace=True,
                   header=None,
                   usecols=my_cols,
                   names=feature_names
                   )

In [317]:
X_test = read_fwf("C:/Users/ejones33/Documents/data/UCI HAR Dataset/test/X_test.txt",
                   #sep=" ",
                   skipinitialspace=True,
                   header=None,
                   usecols=my_cols,
                   names=feature_names
                   )

In [318]:
X_train.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288585          -0.020294          -0.132905         -0.995279   
1           0.278419          -0.016411          -0.123520         -0.998245   
2           0.279653          -0.019467          -0.113462         -0.995380   
3           0.279174          -0.026201          -0.123283         -0.996091   
4           0.276629          -0.016570          -0.115362         -0.998139   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tGravityAcc-mean()-X  \
0         -0.983111         -0.913526              0.963396   
1         -0.975300         -0.960322              0.966561   
2         -0.967187         -0.978944              0.966878   
3         -0.983403         -0.990675              0.967615   
4         -0.980817         -0.990482              0.968224   

   tGravityAcc-mean()-Y  tGravityAcc-mean()-Z  tGravityAcc-std()-X  \
0             -0.140840              0.115375            -0.985250   
1             -0.141551              0.109379            -0.997411   
2             -0.142010              0.101884            -0.999574   
3             -0.143976              0.099850            -0.996646   
4             -0.148750              0.094486            -0.998429   

              ...              fBodyGyro-std()-Y  fBodyGyro-std()-Z  \
0             ...                      -0.973886          -0.994035   
1             ...                      -0.987168          -0.989785   
2             ...                      -0.993399          -0.987328   
3             ...                      -0.991646          -0.988678   
4             ...                      -0.991956          -0.987944   

   fBodyAccMag-mean()  fBodyAccMag-std()  fBodyBodyAccJerkMag-mean()  \
0           -0.952155          -0.956134                   -0.993726   
1           -0.980857          -0.975866                   -0.990335   
2           -0.987795          -0.989015                   -0.989280   
3           -0.987519          -0.986742                   -0.992769   
4           -0.993591          -0.990063                   -0.995523   

   fBodyBodyAccJerkMag-std()  fBodyBodyGyroMag-mean()  fBodyBodyGyroMag-std()  \
0                  -0.993755                -0.980135               -0.961309   
1                  -0.991960                -0.988296               -0.983322   
2                  -0.990867                -0.989255               -0.986028   
3                  -0.991700                -0.989413               -0.987836   
4                  -0.994389                -0.991433               -0.989059   

   fBodyBodyGyroJerkMag-mean()  fBodyBodyGyroJerkMag-std()  
0                    -0.991990                   -0.990697  
1                    -0.995854                   -0.996399  
2                    -0.995031                   -0.995127  
3                    -0.995221                   -0.995237  
4                    -0.995093                   -0.995465  

[5 rows x 66 columns]

In [319]:
X_train.shape

(7352, 66)

Now we will read in the files containing the activity label (Walking, laying down, etc.) associated with each observation in the datasets. They are initially labeled using integers 1:6, and we wish to rename each observation using descriptive words. 

In [320]:
activity_labels = read_csv('C:/Users/ejones33/Documents/data/UCI HAR Dataset/activity_labels.txt',sep=" ",header=None)
activity_labels

0                   1
0  1             WALKING
1  2    WALKING_UPSTAIRS
2  3  WALKING_DOWNSTAIRS
3  4             SITTING
4  5            STANDING
5  6              LAYING

In [321]:
activity_labels = list(activity_labels[1])

In [322]:
activity = y_train.append(y_test)

In [323]:
activity = activity["activity"].astype('category', categories=range(1,7))
activity.head()

0    5
1    5
2    5
3    5
4    5
Name: activity, dtype: category
Categories (6, int64): [1, 2, 3, 4, 5, 6]

In [324]:
activity.cat.categories = activity_labels
activity = DataFrame(activity, columns=["activity"])
activity.head()

activity
0  STANDING
1  STANDING
2  STANDING
3  STANDING
4  STANDING

Now we put the features together with subject number and activity to create a tidy data set. 

`subject` and `activity` are single-column data frames with 10,299 observations. `X_train` and `X_test` are still separate data frames which will be stacked and then joined to the other columns.

In [325]:
features = X_train.append(X_test)
features.shape

(10299, 66)

In [326]:
tidy_data = DataFrame()

In [327]:
tidy_data = concat([subject,activity,features], axis=1)
tidy_data.shape

(10299, 68)

In [328]:
tidy_data.index = range(10299)
tidy_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10299 entries, 0 to 10298
Data columns (total 68 columns):
subject                        10299 non-null int64
activity                       10299 non-null category
tBodyAcc-mean()-X              10299 non-null float64
tBodyAcc-mean()-Y              10299 non-null float64
tBodyAcc-mean()-Z              10299 non-null float64
tBodyAcc-std()-X               10299 non-null float64
tBodyAcc-std()-Y               10299 non-null float64
tBodyAcc-std()-Z               10299 non-null float64
tGravityAcc-mean()-X           10299 non-null float64
tGravityAcc-mean()-Y           10299 non-null float64
tGravityAcc-mean()-Z           10299 non-null float64
tGravityAcc-std()-X            10299 non-null float64
tGravityAcc-std()-Y            10299 non-null float64
tGravityAcc-std()-Z            10299 non-null float64
tBodyAccJerk-mean()-X          10299 non-null float64
tBodyAccJerk-mean()-Y          10299 non-null float64
tBodyAccJerk-mean()-Z       

The `DataFrame.info()` function shows a summary of the table. 

We can see that we have each observation categorized by subject and activity with their relative measurements. 

The final step is to combine all measurements for each subject and activity into an average value. For example, let's see how many observations were taken of Subject 12 lying down: 

In [329]:
len(tidy_data[(tidy_data.subject == 12) & (tidy_data.activity == "LAYING")])

60

Our goal is to take the average of every subject (1-30) and every activity, so our final table will have 30*6=180 rows. 

We also need to convert `tidy_data["subject"]` to categories in order to create our two-level combinations. 

In [330]:
tidy_data["subject"] = tidy_data["subject"].astype("category", range(1,31))

In [331]:
by_subj_act = tidy_data.groupby(["subject","activity"])
by_subj_act

In [332]:
by_subj_act.mean().reset_index()

subject            activity  tBodyAcc-mean()-X  tBodyAcc-mean()-Y  \
0         1             WALKING           0.277331          -0.017384   
1         1    WALKING_UPSTAIRS           0.255462          -0.023953   
2         1  WALKING_DOWNSTAIRS           0.289188          -0.009919   
3         1             SITTING           0.261238          -0.001308   
4         1            STANDING           0.278918          -0.016138   
5         1              LAYING           0.221598          -0.040514   
6         2             WALKING           0.276427          -0.018595   
7         2    WALKING_UPSTAIRS           0.247165          -0.021412   
8         2  WALKING_DOWNSTAIRS           0.277615          -0.022661   
9         2             SITTING           0.277087          -0.015688   
10        2            STANDING           0.277911          -0.018421   
11        2              LAYING           0.281373          -0.018159   
12        3             WALKING           0.275567          -0.017177   
13        3    WALKING_UPSTAIRS           0.260820          -0.032411   
14        3  WALKING_DOWNSTAIRS           0.292423          -0.019355   
15        3             SITTING           0.257198          -0.003503   
16        3            STANDING           0.280047          -0.014338   
17        3              LAYING           0.275517          -0.018956   
18        4             WALKING           0.278582          -0.014840   
19        4    WALKING_UPSTAIRS           0.270877          -0.031980   
20        4  WALKING_DOWNSTAIRS           0.279965          -0.009802   
21        4             SITTING           0.271538          -0.007163   
22        4            STANDING           0.280500          -0.009489   
23        4              LAYING           0.273572          -0.015003   
24        5             WALKING           0.277842          -0.017285   
25        5    WALKING_UPSTAIRS           0.268459          -0.032527   
26        5  WALKING_DOWNSTAIRS           0.293544          -0.008501   
27        5             SITTING           0.273694          -0.009901   
28        5            STANDING           0.282544          -0.007004   
29        5              LAYING           0.278334          -0.018304   
..      ...                 ...                ...                ...   
150      26             WALKING           0.279264          -0.015426   
151      26    WALKING_UPSTAIRS           0.272691          -0.028163   
152      26  WALKING_DOWNSTAIRS           0.279285          -0.012626   
153      26             SITTING           0.258244          -0.007134   
154      26            STANDING           0.281127          -0.016661   
155      26              LAYING           0.271646          -0.019190   
156      27             WALKING           0.276849          -0.016648   
157      27    WALKING_UPSTAIRS           0.265770          -0.020095   
158      27  WALKING_DOWNSTAIRS           0.297544          -0.013556   
159      27             SITTING           0.273941          -0.015527   
160      27            STANDING           0.279567          -0.016593   
161      27              LAYING           0.274102          -0.017987   
162      28             WALKING           0.281228          -0.015676   
163      28    WALKING_UPSTAIRS           0.262006          -0.027944   
164      28  WALKING_DOWNSTAIRS           0.293642          -0.022023   
165      28             SITTING           0.276978          -0.018540   
166      28            STANDING           0.277795          -0.017264   
167      28              LAYING           0.275913          -0.016754   
168      29             WALKING           0.272000          -0.016292   
169      29    WALKING_UPSTAIRS           0.265423          -0.029947   
170      29  WALKING_DOWNSTAIRS           0.293140          -0.014941   
171      29             SITTING           0.277180          -0.016631   
172      29            STANDING           0.277965          -0

Ta-da! That was the end result of the Coursera data project. I think it was much more efficient in pandas. The R code is attached below if you wish to compare. I'm sure it's not as efficient as it could be, but the pandas/python syntax is much more intuitive to me, and more consistent than the R version. 

```R
# Coursera Data Science: Getting and Cleaning Data
# Course Project

# Categorize the rows by subject and activity
subject_train <- read.table("./data/UCI HAR Dataset/train/subject_train.txt",col.names="Subject")
n_train <- dim(subject_train)[1]
subject_test <- read.table("./data/UCI HAR Dataset/test/subject_test.txt",col.names="Subject")
n_test <- dim(subject_test)[1]
# dim(subject_train) # 7352 by 1
# dim(subject_test) # 2947 1
subject <- rbind(data.frame(subject_train),data.frame(subject_test))
levels(subject) <- c(1:30)

# Rename the elements of y_test and y_train to be activity names using plyr's mapvalues
library(plyr); library(data.table); library(dplyr)
activity_labels <- read.table("./data/UCI HAR Dataset/activity_labels.txt")
activities <- factor(activity_labels[,2])

y_train <- read.table("./data/UCI HAR Dataset/train/y_train.txt")
y_test <- read.table("./data/UCI HAR Dataset/test/y_test.txt")
y_all <- rbind(y_train,y_test)
activity <- mapvalues(y_all[,1], from=c(1:6), to=as.character(activities))
activity <- data.frame(activity)
names(activity) <- "Activity"

# Create a vector for denoting whether a row is testing or training data
# data_type <- data.frame(c(rep("TRAIN",times=n_train),rep("TEST",times=n_test)))
# names(data_type) <- "Data_Type"

# Begin to assemble the tidy data frame by column binding
# col.names <- c("Test/Train", "Subject", "Activity")
# tidy_data <- data.frame(data_type, subject, activity)
tidy_data <- data.frame(subject, activity)
head(tidy_data)

# Extract mean and std measurements from features
features <- read.table("./data/UCI HAR Dataset/features.txt")
# dim(features) # 561 by 2
# tBodyAcc-mean(), -std() are [1:6]
# tGravityAcc-mean(), -std() are [41:46]
# tBodyAccJerk-mean(), -std() are [81:86]
# and so on 

# Pull out the 66 variables that pertain to mean or std, but not meanFreq
feature_names <- c(make.names(features[,2]))
substring <- ".mean..|.std.."
is_mean_std <- grepl(substring, feature_names)
is_mean_freq <- grep("meanFreq", feature_names)
is_mean_std[is_mean_freq] <- FALSE
feature_names <- feature_names[is_mean_std]
# Make names syntactically valid
feature_names <- gsub("...", ".", feature_names,fixed=TRUE)
feature_names <- gsub("..", "", feature_names,fixed=TRUE)

X_train <- read.table("./data/UCI HAR Dataset/train/X_train.txt")
X_test <- read.table("./data/UCI HAR Dataset/test/X_test.txt")
X_all <- rbind(X_train,X_test)
X_mean_std <- X_all[,is_mean_std]
# dim(X_all) # 10299 by 561
# dim(X_mean_std) # 10299 by 66
names(X_mean_std) <- feature_names
tidy_data <- cbind(tidy_data, X_mean_std)
tidy_data <- arrange(tidy_data, Subject, Activity)
# 10299 obs. of 68 variables

# Second, independent data set with averages by the 6 activities and 30 subjects
by_subj_act <- group_by(tidy_data, Subject, Activity)
tidy_data2 <- summarise_each(by_subj_act, funs(mean))
# 180 obs. of 68 variables
# head(tidy_data2,n=10)
```